File: app/src/main/java/com/example/lockit_mobile_app/Signup.java
package com.example.lockit_mobile_app;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.FormBody;
import okhttp3.OkHttp;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class Signup extends AppCompatActivity {

    OkHttpClient client;
    EditText editTextUsername,editTextEmail,editTextPhoneNumber,editTextPassword,editTextRetypePassword;
    Button buttonSignup;
    ProgressBar loding;
    TextView login;
    String getURL = "";
    String postUrl = "http://10.0.2.2:5001/lockit-332b1/us-central1/app/register";

    @SuppressLint("MissingInflatedId")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_signup);
        client = new OkHttpClient();
        editTextUsername = findViewById(R.id.username);
        editTextEmail = findViewById(R.id.email);
        editTextPhoneNumber = findViewById(R.id.phonenumber);
        editTextPassword = findViewById(R.id.password);
        editTextRetypePassword = findViewById(R.id.retypepassword);
        buttonSignup = findViewById(R.id.signup);
        loding = findViewById(R.id.progressBar);
        login = findViewById(R.id.loginText);

        login.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(getApplicationContext(),Login.class);
                startActivity(intent);
                finish();
            }
        });
        buttonSignup.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                loding.setVisibility(View.VISIBLE);
                String username,email,phonNumber,password,retypePassword;
                username = String.valueOf(editTextUsername.getText());
                email = String.valueOf(editTextEmail.getText());
                phonNumber = String.valueOf(editTextPhoneNumber.getText());
                password = String.valueOf(editTextPassword.getText());
                retypePassword = String.valueOf(editTextRetypePassword.getText());

                if (!isNetworkAvailable()) {
                    Toast.makeText(Signup.this, "No internet connection", Toast.LENGTH_SHORT).show();
                    return;
                }

                if (TextUtils.isEmpty(username)){
                    Toast.makeText(Signup.this, "Username can't be empty", Toast.LENGTH_SHORT).show();
                    return;
                }
                if (TextUtils.isEmpty(email) || !isValidEmail(email)) {
                    Toast.makeText(Signup.this, "Enter a valid email address", Toast.LENGTH_SHORT).show();
                    return;
                }
                if (TextUtils.isEmpty(phonNumber)){
                    Toast.makeText(Signup.this, "Phone Number can't be empty", Toast.LENGTH_SHORT).show();
                    return;
                }
                if (TextUtils.isEmpty(password)){
                    Toast.makeText(Signup.this, "Password can't be empty", Toast.LENGTH_SHORT).show();
                    return;
                }
                if (TextUtils.isEmpty(retypePassword)){
                    Toast.makeText(Signup.this, "Retype password can't be empty", Toast.LENGTH_SHORT).show();
                    return;
                }
                if (!TextUtils.equals(password,retypePassword)){
                    Toast.makeText(Signup.this, "Passwords are not matching", Toast.LENGTH_SHORT).show();
                    return;
                }

                signup(username,email,phonNumber,password);
            }
        });

    }

    private boolean isNetworkAvailable() {
        ConnectivityManager connectivityManager = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
        return activeNetworkInfo != null && activeNetworkInfo.isConnected();
    }

    private boolean isValidEmail(String email) {
        String emailRegex = "^[a-zA-Z0-9_+&*-]+(?:\\.[a-zA-Z0-9_+&*-]+)*@(?:[a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,7}$";
        return email.matches(emailRegex);
    }

    public void signup(String username,String email,String phonNumber,String password){
        RequestBody requestBody = new FormBody.Builder()
                .add("username",username)
                .add("email",email)
                .add("phoneNumber",phonNumber)
                .add("password",password)
                .build();
        Request request = new Request.Builder().url(postUrl).post(requestBody).build();
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                e.printStackTrace();
                loding.setVisibility(View.GONE);
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        Toast.makeText(Signup.this, "Something went wrong", Toast.LENGTH_SHORT).show();
                    }
                });
            }
            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
                loding.setVisibility(View.GONE);
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        try {
                            Toast.makeText(Signup.this, response.body().string(), Toast.LENGTH_SHORT).show();
                        } catch (IOException e) {
                            throw new RuntimeException(e);
                        }
                    }
                });
            }

        });
    }
}

File: app/src/main/java/com/example/lockit_mobile_app/SignUp.kt
package com.example.lockit_mobile_app

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkRequest
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import okhttp3.Call
import okhttp3.Callback
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okio.IOException
import org.json.JSONObject

class SignUp : AppCompatActivity()  {

    private lateinit var client: OkHttpClient

    private lateinit var editTextUsername: EditText
    private lateinit var editTextEmail: EditText
    private lateinit var editTextPhoneNumber: EditText
    private lateinit var editTextPassword: EditText
    private lateinit var editTextRetypePassword: EditText

    private lateinit var buttonSignup: Button

    private lateinit var loading: ProgressBar

    private lateinit var login: TextView

    private val postURL: String = "https://lockit-backend-api.onrender.com/api/users/register"

    val loadingAlert = LoadingAlert(this)

     @SuppressLint("WrongViewCast", "MissingInflatedId")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.signup)

        client = OkHttpClient()

        editTextUsername = findViewById(R.id.username)
        editTextEmail = findViewById(R.id.email)
        editTextPhoneNumber = findViewById(R.id.phonenumber)
        editTextPassword = findViewById(R.id.password)
        editTextRetypePassword = findViewById(R.id.retypepassword)

        buttonSignup = findViewById(R.id.signup)

        loading = findViewById(R.id.progressBar)

        login = findViewById(R.id.login)

        login.setOnClickListener{
            val intent = Intent(applicationContext,Login::class.java)
            startActivity(intent)
            finish()
        }

        buttonSignup.setOnClickListener {
            loading.visibility = View.VISIBLE
            val username = editTextUsername.text.toString()
            val email = editTextEmail.text.toString()
            val phoneNumber = editTextPhoneNumber.text.toString()
            val password = editTextPassword.text.toString()
            val retypePassword = editTextRetypePassword.text.toString()

            if (!isNetworkAvailable()) {
                Toast.makeText(this@SignUp, "No internet connection", Toast.LENGTH_SHORT).show()

                loading.visibility = View.GONE

                return@setOnClickListener
            }

            if (username.isEmpty()) {
                Toast.makeText(this@SignUp, "Username can't be empty", Toast.LENGTH_SHORT).show()

                loading.visibility = View.GONE

                return@setOnClickListener
            }

            if (email.isEmpty() || !Utils.isValidEmail(email)) {
                Toast.makeText(this@SignUp, "Enter a valid email address", Toast.LENGTH_SHORT).show()

                loading.visibility = View.GONE

                return@setOnClickListener
            }

            if (phoneNumber.isEmpty()) {
                Toast.makeText(this@SignUp, "Phone Number can't be empty", Toast.LENGTH_SHORT).show()

                loading.visibility = View.GONE

                return@setOnClickListener
            }

            if (password.isEmpty()) {
                Toast.makeText(this@SignUp, "Password can't be empty", Toast.LENGTH_SHORT).show()

                loading.visibility = View.GONE

                return@setOnClickListener
            }

            if (retypePassword.isEmpty()) {
                Toast.makeText(this@SignUp, "Retype password can't be empty", Toast.LENGTH_SHORT).show()

                loading.visibility = View.GONE

                return@setOnClickListener
            }

            if (password != retypePassword) {
                Toast.makeText(this@SignUp, "Passwords are not matching", Toast.LENGTH_SHORT).show()

                loading.visibility = View.GONE

                return@setOnClickListener
            }
            loadingAlert.startLoading()
            signup(username, email, phoneNumber, password)
        }

    }
    private fun isNetworkAvailable(): Boolean {
        val connectivityManager = getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

        val networkCallback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) {
                // Network is available
//                Toast.makeText(this@SignUp, "Online", Toast.LENGTH_SHORT).show()
            }

            override fun onLost(network: Network) {
                // Network is lost
                Toast.makeText(this@SignUp, "Offline", Toast.LENGTH_SHORT).show()
            }
        }

        val networkRequest = NetworkRequest.Builder().build()

        connectivityManager.registerNetworkCallback(networkRequest, networkCallback)

        val capabilities = connectivityManager.getNetworkCapabilities(connectivityManager.activeNetwork)
        val isConnected = capabilities != null

        connectivityManager.unregisterNetworkCallback(networkCallback)

        return isConnected
    }

    private fun signup(username: String, email: String, phoneNumber: String, password: String) {
        val requestBody = FormBody.Builder()
            .add("username", username)
            .add("email", email)
            .add("phoneNumber", phoneNumber)
            .add("password", password)
            .build()

        val request = Request.Builder().url(postURL).post(requestBody).build()

        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                e.printStackTrace()
                runOnUiThread {
                    loadingAlert.stopLoading()
                    Toast.makeText(this@SignUp, "Something went wrong", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onResponse(call: Call, response: Response) {
                runOnUiThread {
                    try {
                        val jsonResponse = response.body?.string()
                        Log.d("Response", jsonResponse ?: "Empty response")

                        // Convert JSON response string to JSONObject
                        val jsonObject = JSONObject(jsonResponse)

                        Toast.makeText(this@SignUp,jsonObject.getString("message"), Toast.LENGTH_SHORT).show()
                        loadingAlert.stopLoading()
                        val intent = Intent(applicationContext,Login::class.java)
                        startActivity(intent)
                        finish()
                    } catch (e: IOException) {
                        throw RuntimeException(e)
                        loadingAlert.stopLoading()
                    }
                }
            }
        })
    }

}
